### PR TITLE
add `inspec.profile.file(...)` for profile files

### PR DESCRIPTION
**Pull request by arlimus**
Heavily depends on the outcome of the RFC discussion here: Fixes https://github.com/chef/inspec/issues/1396

WIP:

* [x] add docs

Fixes https://github.com/chef/inspec/issues/1396

### CODE REVIEW

File: docs/profiles.md
Comment (adamleff) [inline docs/profiles.md]: `s/coveres/covers`

Comment (adamleff) [reply docs/profiles.md]: This feels like a semi-bad example of why you'd want to use a file because something like a port list is easy-enough to put in the control itself.

Maybe something like "This allows a profile to include a file that contains many different fields and values that could be used through the profile in across mulitple control files without including that data in each control file."

File: docs/profiles.md
Comment (adamleff) [inline docs/profiles.md]: If you decide to change the wording, maybe the example could be a YAML file like so:

```yaml
  web_alpha:
    service_name: httpd-alpha
    port: 80
  web_beta:
    service_name: httpd-beta
    port: 8080
```

... and the test could be:

```ruby
services = YAML.parse(inspec.profile.file('services.yml'))

services.each do |svc|
  describe service(svc['name']) do
    it { should be_running }
  end
  describe port(svc['port']) do
    it { should be_listening }
  end
end
```

Sorry if I'm being pedantic, but I wouldn't want to encourage profiles becoming too abstracted and difficult to follow. I'd want the new `file` functionality to be used when it's really truly necessary.

File: lib/source_readers/flat.rb
Comment (adamleff) [inline lib/source_readers/flat.rb]: I think it's going to get super-confusing for newcomers to this codebase that we call this "files" but it only includes files in the `files` directory, when the controls are stored in files also.

Should we call them "supplemental_files" or "data_files" or something of that nature?

File: lib/inspec/runtime_profile.rb
Comment (adamleff) [inline lib/inspec/runtime_profile.rb]: Do we need to prepopulate this hash with the files AND their contents? The SourceReader is already doing that, and then we do it again here just to have different keys... feels like we're consuming 2x the memory we need for files that may not even get used by the controls.

It feels like we should let the SourceReader provide the list of files and provide a method to return the contents, and all this class has to do is take a parameter, prefix it with "files/" and then pass it to the SourceReader to get us the content... and to list the files, it can strip "files/" before it hands it back to the caller.

Comment (arlimus) [reply lib/inspec/runtime_profile.rb]: 👍 will implement this thank you!

File: lib/source_readers/flat.rb
Comment (arlimus) [reply lib/source_readers/flat.rb]: 👍 yeah had a similar feeling, just not a good name. i'll go with `data_files`, can't thing of anything else and it looks good

File: docs/profiles.md
Comment (arlimus) [reply docs/profiles.md]: Funny enough, I have taken the ports exactly from a customer example 😁 I also see why they want to use it that way and it's ok from my perspective.

I also agree with your premise! I wanted to avoid parsing a file to keep it really simple, but it might not hurt that much. I'll take your example, thank you for sharing.

Comment (adamleff) [reply docs/profiles.md]: Totally your call. I'd be pretty frustrated if I came across a profile that had values like that in a separate file... just feels like unnecessary abstraction.

If those values were used in 3+ different control files, then sign me up, it sounds great. If it's just in one, I don't understand the benefit.

File: docs/profiles.md
Comment (chris-rock) [inline docs/profiles.md]: I assume this is `services.yml` now?

File: docs/profiles.md
Comment (chris-rock) [inline docs/profiles.md]: should we use:
```
my_services = yaml(content: inspec.profile.file('services.yml'))
```

**Review by chris-rock — COMMENTED** [review]
How is that working with inherited profiles? Can I access files from inherited profiles?

Will a control still work, if we use it in combination with `include_controls`?

File: docs/profiles.md
Comment (arlimus) [reply docs/profiles.md]: ah fail... thank you!

**Comment by arlimus** [comment]
Good points on inheritance @chris-rock . Short answer: It should work, down the line. It doesn't work, right now. The reason is the source-reader and how it handles dependencies (at least in a couple of corner-cases I tested). Love to address this separately.

**Comment by chris-rock** [comment]
Thank you @arlimus for clarification. This should not block the feature, we just need to be aware of the limitation.

**Review by chris-rock — APPROVED** [review]
Thank you for the quick turnaround @arlimus

File: docs/profiles.md
Comment (arlimus) [reply docs/profiles.md]: great point!

**Comment by aaronlippold** [comment]
This would also cover use cases where we need to specify things like : approved_accounts or approved_x data structure for login accounts or debase roles, valid or invalid connection types, etc. In the end, basically any data structure you could want to test, or file to compare etc. Good add.

**Comment by aaronlippold** [comment]
Given https://github.com/CrunchyData/postgresql-baseline/issues/9 as a use case, one thing that we may want to support with this is the ability to also pass in a file on the command line or during a run. 

I can also see this location being a variable that may come from chef or automate or even gitlab.

Simple use:
1. I have an {mydb}-approved_account_exceptions.yml file that I pass to the general DB profile to use
2. {mydb} changes as do the approved exceptions from DB to DB

Advanced Use:
In Automate or in the larger CI/CD pipeline - from env to env - the required support files may have a standard format but - using @arlimus examples - the services file and ports may change between environments.

**Comment by aaronlippold** [comment]
@arlimus @chris-rock @adamleff I love this new option, one question, what happens when I want to create a file in the files directory called 'approved users and connections' - the YAML file has a standard structure and I want users to fill in this file so that the generalized control I wrote in the profile will run with respect to their target system. I.e. the file name will stay the same, the structure, only the data changes relative to the system it is being run on, would you say then that we would use attributes to over-ride that data or would you want to pass in the new YAML file?

**Comment by arlimus** [comment]
We had a small chat about ^^ on the phone: Some aspects may be handled via profile files, usually those that are universal and shared with the profile itself. Things that are less universal are recommended to be handled via attributes.

**Pull request merged**
